fix: Apply allocation filters before aggregation to ensure AND operator order independence

spa-raj opened the pull request:
## What does this PR change?
* The AND operator (+) in v2 allocation filters was order-dependent, causing filters like `namespace:"ns1"+cluster:"non-existent"` to behave differently than `cluster:"non-existent"+namespace:"ns1"`.

* Root cause: Aggregation was happening before filtering, which could merge or lose allocation properties (like cluster), making them unavailable for filter matching. This caused the first filter condition to match while subsequent conditions had incomplete data.

* Solution: Reordered operations to filter -> aggregate -> accumulate. Filtering now happens on individual allocations with all properties intact, ensuring consistent AND behavior regardless of filter order. 

## Does this PR relate to any other PRs?
* No

## How will this PR impact users?
*  Users experiencing incorrect results with AND filters (+) in v2 allocation API will now get correct behavior. Previously, filters like namespace:"ns1"+cluster:"non-existent" could incorrectly return data, while reversing the order would correctly return nothing. 
* After this fix, AND filters work consistently regardless of order - both conditions must match for allocations to be included. Users who may have worked around this bug by ordering their filters specifically will need to update their queries.

## Does this PR address any GitHub or Zendesk issues?
* Fixes #3371 

## How was this PR tested?
* Added unit test Test_AllocationFilterAnd_OrderIndependence in core/pkg/opencost/allocationfilter_test.go:745-820 that validates:
  - AND filters produce identical results regardless of filter order
  - The specific bug scenario (namespace matches, cluster doesn't) correctly returns false
  - Both matching conditions return true
  - All allocation properties remain available for filtering before aggregation

## Does this PR require changes to documentation?
* No

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
* This should be labeled for next release but I don't have the access to label it.

spa-raj committed:
fix: Apply allocation filters before aggregation to ensure AND operator order independence

The AND operator (+) in v2 allocation filters was order-dependent, causing
filters like `namespace:"ns1"+cluster:"non-existent"` to behave differently
than `cluster:"non-existent"+namespace:"ns1"`.

Root cause: Aggregation was happening before filtering, which could merge or
lose allocation properties (like cluster), making them unavailable for filter
matching. This caused the first filter condition to match while subsequent
conditions had incomplete data.

Solution: Reordered operations to filter → aggregate → accumulate. Filtering
now happens on individual allocations with all properties intact, ensuring
consistent AND behavior regardless of filter order.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Merge branch 'opencost:develop' into bug/v2-filter-AND-parameter

ameijer commented on pkg/costmodel/aggregation.go:
```diff
@@ -227,7 +227,11 @@ func (a *Accesses) ComputeAllocationHandler(w http.ResponseWriter, r *http.Reque
+	// on individual allocations before they are aggregated. This ensures filters
+	// can match on all allocation properties (like cluster, node, etc.) before
+	// they are potentially lost or merged during aggregation.
+	asr, err := a.Model.QueryAllocation(window, step, nil, includeIdle, idleByNode, includeProportionalAssetResourceCosts, includeAggregatedMetadata, sharedLoadBalancer, opencost.AccumulateOptionNone, shareIdle)
```
wow. yeah. the fact that we don't even pass in the filter here is a red flag. Your approach is right, I just wanted to ask why not do this in the QueryAllocation method, instead of further down this method? Just asking, not demanding that we do that

spa-raj replied:
I kept it in the handler to avoid changing the core method signature, but you are right that's the wrong trade-off. I'm already duplicating the same filtering logic in both handlers,  and we're doing unnecessary work (computing idle allocations for stuff we just filter out anyway).

I should add a filter string param to QueryAllocation and do the filtering there before aggregation. Existing callers can just pass "".

Want me to refactor it that way?

ameijer replied:
yes let's do that!

spa-raj committed:
refactor: move allocation filtering into QueryAllocation method

Centralize allocation filter logic in QueryAllocation to eliminate
duplication and improve performance by filtering before aggregation.

- Add filterString parameter to QueryAllocation
- Remove duplicate filtering from ComputeAllocationHandler
- Filter allocations before computing idle costs (performance win)

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Merge branch 'opencost:develop' into bug/v2-filter-AND-parameter

spa-raj committed:
Merge branch 'opencost:develop' into bug/v2-filter-AND-parameter
